src/shim/state_tree.rs

```diff
@@ -291,6 +291,36 @@ where
             StateTree::V0(_) => bail!("StateTree::set_actor not supported on old state trees"),
         }
     }
+
+    /// Returns the public key type of
+    /// address(`BLS`/`SECP256K1`) of an actor identified by `addr`,
+    /// or its delegated address.
+    pub fn resolve_to_deterministic_addr(
+        &self,
+        store: &impl Blockstore,
+        addr: Address,
+    ) -> anyhow::Result<Address> {
+        use crate::shim::address::Protocol::*;
+        match addr.protocol() {
+            BLS | Secp256k1 | Delegated => Ok(addr),
+            _ => {
+                let actor = self
+                    .get_actor(&addr)?
+                    .with_context(|| format!("failed to find actor: {addr}"))?;
+
+                // if state.Version() >= types.StateTreeVersion5
+                if matches!(self, Self::FvmV4(_) | Self::FvmV3(_)) {
+                    if let Some(address) = actor.delegated_address {
+                        return Ok(address.into());
+                    }
```

Comment:
Do we even need to check if it's FVM4 or FVM3? What if (when) we get FVM5?

Reply:
That's good point, it's better to check older versions

Reply:
Fixed

Reply:
But do we need to have this check at all? I mean, paranoid checks are fine by me, but I believe the `delegated_address` should be `None` for older versions.

Reply:
> But do we need to have this check at all? I mean, paranoid checks are fine by me, but I believe the `delegated_address` should be `None` for older versions.

I think you are right, but I'm inclined to keep the logic the same with Go, what do you think?

```go
if state.Version() >= types.StateTreeVersion5 {
		if act.Address != nil {
			// If there _is_ an f4 address, return it as "key" address
			return *act.Address, nil
		}
	}
```

Reply:
Sure, it's a cheap operation anyway.